{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\statespace\\sarimax.py:966: UserWarning: Non-stationary starting autoregressive parameters found. Using zeros as starting parameters.\n",
      "  warn('Non-stationary starting autoregressive parameters'\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\statespace\\sarimax.py:978: UserWarning: Non-invertible starting MA parameters found. Using zeros as starting parameters.\n",
      "  warn('Non-invertible starting MA parameters found.'\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\statespace\\sarimax.py:978: UserWarning: Non-invertible starting MA parameters found. Using zeros as starting parameters.\n",
      "  warn('Non-invertible starting MA parameters found.'\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\tsa\\base\\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency H will be used.\n",
      "  self._init_dates(dates, freq)\n",
      "c:\\Users\\johnl\\anaconda3\\envs\\my_python_env\\lib\\site-packages\\statsmodels\\base\\model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals\n",
      "  warnings.warn(\"Maximum Likelihood optimization failed to \"\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from statsmodels.tsa.statespace.sarimax import SARIMAX\n",
    "from statsmodels.tsa.arima.model import ARIMA\n",
    "from sklearn.metrics import mean_absolute_error, mean_squared_error\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import product\n",
    "\n",
    "data = pd.read_csv('./CFS_based_feature.csv')\n",
    "data['timestamp'] = pd.to_datetime(data['timestamp'])  # Ensure timestamp is datetime\n",
    "\n",
    "SEED = 12\n",
    "np.random.seed(SEED)\n",
    "\n",
    "# Set timestamp as index\n",
    "data.set_index('timestamp', inplace=True)\n",
    "\n",
    "# Define target and exogenous features\n",
    "target = data['cooling_power']\n",
    "exog = data.drop(columns=['cooling_power'])\n",
    "\n",
    "# Split the dataset into train, validation, and test sets\n",
    "train_size = 0.7\n",
    "val_size = 0.15\n",
    "\n",
    "train_end = int(len(data) * train_size)\n",
    "val_end = train_end + int(len(data) * val_size)\n",
    "\n",
    "y_train, y_val, y_test = target[:train_end], target[train_end:val_end], target[val_end:]\n",
    "exog_train, exog_val, exog_test = exog[:train_end], exog[train_end:val_end], exog[val_end:]\n",
    "\n",
    "best_arimax_order = None\n",
    "best_arimax_aic = float('inf')\n",
    "\n",
    "p_values = range(1, 3)\n",
    "d_values = range(1, 2)\n",
    "q_values = range(1, 3)\n",
    "\n",
    "for order in product(p_values, d_values, q_values):\n",
    "    try:\n",
    "        model = ARIMA(endog=y_train, exog=exog_train, order=order)\n",
    "        model_fit = model.fit()\n",
    "        if model_fit.aic < best_arimax_aic:\n",
    "            best_arimax_aic = model_fit.aic\n",
    "            best_arimax_order = order\n",
    "    except Exception as e:\n",
    "        continue\n",
    "\n",
    "# Hyperparameter tuning for SARIMAX using AIC\n",
    "best_sarimax_order = None\n",
    "best_sarimax_seasonal_order = None\n",
    "best_sarimax_aic = float('inf')\n",
    "\n",
    "seasonal_p_values = range(0, 2)\n",
    "seasonal_d_values = range(0, 2)\n",
    "seasonal_q_values = range(0, 2)\n",
    "seasonal_m = [24]\n",
    "\n",
    "for order in product(p_values, d_values, q_values):\n",
    "    for seasonal_order in product(seasonal_p_values, seasonal_d_values, seasonal_q_values, seasonal_m):\n",
    "        try:\n",
    "            model = SARIMAX(endog=y_train, exog=exog_train, order=order, seasonal_order=seasonal_order)\n",
    "            model_fit = model.fit(disp=False)\n",
    "            if model_fit.aic < best_sarimax_aic:\n",
    "                best_sarimax_aic = model_fit.aic\n",
    "                best_sarimax_order = order\n",
    "                best_sarimax_seasonal_order = seasonal_order\n",
    "        except Exception as e:\n",
    "            continue\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ariam_true=0\n",
    "if ariam_true==0:\n",
    "    final_model = ARIMA(endog=pd.concat([y_train,y_val],axis=0), exog=pd.concat([exog_train,exog_val],axis=0), order=best_arimax_order)\n",
    "    model_type = 'ARIMAX'\n",
    "else:\n",
    "    final_model = SARIMAX(endog=y_train, exog=exog_train, order=best_sarimax_order, seasonal_order=best_sarimax_seasonal_order)\n",
    "    model_type = 'SARIMAX'\n",
    "\n",
    "final_model_fit = final_model.fit()\n",
    "#print(exog_test.shape, )\n",
    "predictions_test = final_model_fit.forecast(steps=len(y_test), exog=exog_test)\n",
    "# exog_test_aligned = exog_test.loc[y_test.index]\n",
    "# predictions_test = final_model_fit.predict(\n",
    "#     start=y_test.index[0],  # Start from the first timestamp in test data\n",
    "#     end=y_test.index[-1],   # End at the last timestamp in test data\n",
    "#     exog=exog_test_aligned\n",
    "# )\n",
    "#print(predictions_test)\n",
    "\n",
    "predictions_test_1=predictions_test.reset_index(drop=True)\n",
    "#print(y_test)\n",
    "y_test_1=y_test.reset_index(drop=True)\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.plot(y_test_1[:200].values, marker='o', label='Ground Truth')\n",
    "plt.plot(predictions_test_1[:200], marker='x', label='Predicted')\n",
    "plt.title(f'CFS - Ground Truth vs Predicted Cooling Power ({model_type})')\n",
    "plt.xlabel('Time Index')\n",
    "plt.ylabel('Cooling Power')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "\n",
    "# Evaluate final model on test set\n",
    "test_mae = mean_absolute_error(y_test, predictions_test)\n",
    "test_rmse = np.sqrt(mean_squared_error(y_test, predictions_test))\n",
    "test_mape = np.mean(np.abs((y_test - predictions_test) / y_test)) * 100\n",
    "\n",
    "print(f\"Test MAE: {test_mae:.2f}\")\n",
    "print(f\"Test RMSE: {test_rmse:.2f}\")\n",
    "print(f\"Test MAPE: {test_mape:.2f}%\")\n",
    "\n",
    "print(best_arimax_order,best_sarimax_order,best_sarimax_seasonal_order)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "my_python_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
